torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,540 @@
+            return list_arg.args[0]
+    return None
+
+def _can_reuse_tensor_memory(self_arg, *, view_of_self: Optional[torch.fx.node.Node] = None):
+    nodes_used = self_arg.meta['node_usages'] if 'node_usages' in self_arg.meta else []
+    # If the current tensor argument is used as an input to any ops later in the graph,
+    # then we can't re-use its memory
+    if any(_ for n_idx, n in nodes_used if n_idx > idx):
+        return False
+    # Special case; if the current node is base, and we're trying to re-inplace
+    # an operator to v where v = view_op(base, args...),
+    # Then it is safe to re-inplace v, which will directly modify, base,
+    #     **as long as** either:
+    # (1) base is also not used anywhere later in the program
+    # (2) base used exactly once later in the program, in the following operation:
+    #     base_updated = view_inverse_op(base, v_updated, args...)
+    # where "args..." must match up exactly, and "view_inverse_op"
+    # must be the "inverse" of view, according to the inverse rules
+    # in the functionalization pass.
+    maybe_base_of_self = _maybe_get_base_of_curr_node(self_arg)
+    if maybe_base_of_self is not None:
+        return _can_reuse_tensor_memory(base_of_self, view_of_self=self_arg)
+    return True
+
+_VIEW_INVERSE_MAP = {
+    torch.ops.aten.diagonal.default: torch.ops.aten.diagonal_scatter.default,
+    torch.ops.aten.select.int: torch.ops.aten.select_scatter.default,
+    torch.ops.aten.slice.Tensor: torch.ops.aten.slice_scatter.default,
+    torch.ops.aten.as_strided.default: torch.ops.aten.as_strided_scatter.default,
+}
+
+# Uses the 'view_of' metadata set by FunctionalizationMetadataProp
+# to determine if two different nodes are identical views of the same base.
+# For example:
+#    a1 = x.view(2, 4)
+#    a2 = a1[0]
+#    b1 = x.view(2, 4)
+#    b2 = b1[0]
+# Here, b2 and a2 are matching view tensors, because their chain of views is identical.
+def _are_matching_view_tensors(view1: torch.fx.node.Node, view2: torch.fx.node.Node) -> bool:
+    base1 = view1.meta.get('view_of', None)
+    base2 = view2.meta.get('view_of', None)
+    if base1 is None and base2 is None:
+        return True
+    if base1 is None or base2 is None:
+        return False
+    # Check that "args..." are the same too.
+    assert len(view1.args) == len(view2.args)
+    assert len(view1.kwargs) == len(view2.kwargs)
+    # If the two views were created through the same view op (with the same args),
+    # then recursively check if their bases are also matching views.
+    if view1.target is view2.target:
+        if all(x == y for (x, y) in zip(
+               itertools.chain(view2.args[1:], view2.kwargs),
+               itertools.chain(view1.args[1:], view1.kwargs))):
+            return _are_matching_view_tensors(base1, base2)
+    return False
+
+
+
+def _satisfies_inverse_view(view_inverse_node: torch.fx.node.Node, tensor_aliases: Set[torch.fx.node.Node]):
+    for alias in tensor_aliases:
```

Comment:
Under what circumstances would we return True from this function, AND there are multiple tensor aliases?

Reply:
That could technically happen if you have multiple view tensors that are "identical". But I think that if you were to CSE so that we never bother to construct the same view tensor twice, then it would be the case that there's at most 1 tensor in `tensor_aliases` that can satisfy the criteria here.